### DISCUSSION TITLE
Trying to merge two existing branches (trying to have a dev branch and prod branch)

### DISCUSSION THREAD

**Discussion by rperry99** [community/community#66435] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

So I have a github pages set up for my portfolio. I am trying not to launch my changes directly to my production branch, so I am using master as a dev branch. I am struggling to figure out how this works though. Like how do I sync the changes that are on master, to the production branch?

I feel like this is a simple solution, but something isn't clicking. I tried looking on the forums for this question if it was already asked, but I am not even sure how to ask it. 

I understand how to push changes to master. If I am working on my local file, I will first do `git checkout -b <somebranchname>`. Then I will do my code, then I will `git add .`, `git commit -m <text>`, and then `git push -u origin <somebranchname>`. Then when I go to github, I have my pull request and I make the merge from there. 

Do I `git pull` my master branch locally, then `git push -u origin production`?

**Comment by spenserblack** [comment, marked as answer]
Locally, there are a few ways to do it:
```shell
git checkout production

git pull origin master

# or
git fetch
git merge origin/master  # I like this for being explicit that I want to merge from the remote's master branch

# or
git checkout master
git pull
git checkout production
git merge master

# ... and more!
```

However, the easiest way is probably going to be to create a pull request from master to production ([it looks like you've done that before](https://github.com/rperry99/2023-portfolio/pull/17)). It's nice to have a record of *when* you deployed to production, which you'll have from PRs and their merge commits.

Also, you probably *don't* want to `push -u origin production` from your local `master` branch. `-u` is short for `--set-upstream`, and I doubt you want your local master branch's upstream to be `production`. IIRC this produces a warning when the remote and local branches don't have the same name, anyway. Just for safety, you might want to set up some branch protections or rulesets for your production branch to require PRs for any changes, to help you avoid accidentally pushing to production.

******

A bit off-topic, but you might want to check out Vercel, which can automate preview deployments for PRs and development branches. Sounds like something you might like with your workflow :wink:

**Reply by rperry99** [reply]
Thank you for this detailed answer! 

I figured that pull requests would be the best way for me to do this. What confused me, was when I did the pull request (the one you linked), my production branch said that it was "1 commit ahead of master". So I did another one to try and sync master and production, and now production says it's "1 commit behind master". So now I am confused again. Lol

I thought about using something like Vercel, (I was looking into Digital Ocean specifically), but I thought that was "too much" for what this project is. Honestly, me doing the master vs production branches is more effort than required for what this is. Lol

**Reply by spenserblack** [reply]
Oh, got it. Do you mean seeing "production is X commits ahead of master" on the GitHub website?

Yeah, that will happen if you merge a PR and generate a merge commit -- the merge commit counts towards the count. And then if a merge commit was created when you synched `master` with `production`, then `master` has the master->production commit that put production ahead, and a production->master commit, putting master ahead one commit :laughing:

Personally, I wouldn't worry about the branches being exactly in sync. But you could try `git merge --ff-only --no-commit` to merge without creating a merge commit (I think `--no-commit` is redundant for `--ff-only`, but it's a bit more explicit). Another strategy you could do is create a workflow to deploy to production, if you want to get fancy :laughing: You can adapt [this workflow](https://github.com/spenserblack/actions-tag-to-release/blob/a240d669c41f30c507ae12e06ac26228bfefbd1b/.github/workflows/sync-master.yml) to be triggered on `workflow_dispatch` (when you click a button) to push to production.

As long as production is never ahead of master, I think you can rebase-merge your master->production PRs to keep them in sync. But TBH I don't like rebase merging, since it can rewrite commits, remove GPG signatures, etc., and can cause confusion for git beginners long-term.

**Reply by rperry99** [reply]
Yeah that's exactly what I mean. Currently, my production branch shows the following: 

![image](https://github.com/community/community/assets/24259120/d204380c-82d8-4852-98c3-1a33801913eb)

If this is normal, then I think I am good to go! I am really trying to get better with git. Before All I ever did was push to my own repos for version control, and never did anything with them. Trying to learn a lot more about it, which was why I was trying to do this master / production branch thing. Lol

I mean, it does make sense that master will be ahead of prod at some point. It just feels weird I guess knowing that right now, they _should_ be the same, considering I have no pending changes in master to push to prod. 

If it's not that concerning that they aren't in sync, then I won't worry about it. 

I really appreciate the thorough answer!
